Pylint consider using with

richard-johansson opened the pull request:
## Overview

1. This PR fixes or fixes part of #12912.
2. This PR does the following: Fixes the disabled pylint check "R1732: Consider using 'with' for resource-allocating operations"
<img width="745" alt="bild" src="https://user-images.githubusercontent.com/113174121/206901524-82df040e-3ed1-4ca1-b15b-6d3c1320c95a.png">

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct
<img width="486" alt="bild" src="https://user-images.githubusercontent.com/113174121/206901543-6058c813-b07d-4ba7-91b6-eb138cdc8b5b.png">

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).
- If you need a review or an answer to a question, and don't have permissions to assign people, **leave a comment** like the following: "{{Question/comment}} @{{reviewer_username}} PTAL". Oppiabot will help assign that person for you.
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).

ricjo462 committed:
Commit dure to RuntimeError: The Python development requirements file requirements_dev.txt was changed by the installation script. Please commit the changes. You can get the changes again by running this command: python -m scripts.install_python_dev_dependencies

ricjo462 committed:
RuntimeError: The Python development requirements file requirements_dev.txt was changed by the installation script. Please commit the changes. You can get the changes again by running this command: python -m scripts.install_python_dev_dependencies

ricjo462 committed:
Merge remote-tracking branch 'upstream/develop' into develop

richard-johansson committed:
Merge branch 'oppia:develop' into develop

ricjo462 committed:
Removed 'consider-using-with' to see which linter checks that fail

ricjo462 committed:
Fixed 'core/controllers/profile_test.py:1265:23: R1732: Consider using 'with' for resource-allocating operations' problem

ricjo462 committed:
Fixed 'core/controllers/profile_test.py:1329:23: R1732: Consider using 'with' for resource-allocating operations
'

ricjo462 committed:
Fixed all 'R1732: Consider using 'with' for resource-allocating operations' problems

ricjo462 committed:
Fixed all 'R1732: Consider using 'with' for resource-allocating operations' in this file

ricjo462 committed:
Added pylint ignore since the file needs to be returned open

richard-johansson commented:
> Hi @richard-johansson, can you complete the following:
> 
>     1. The proof that changes are correct has not been provided, please make sure to upload a image/video showing that the changes are correct. Or include a sentence saying "No proof of changes needed because" and the reason why proof of changes cannot be provided.
>        Thanks!

done.

U8NWXD reviewed (changes requested):
@richard-johansson thanks for opening this PR! It looks like there are still a few spots left where the lint check is failing: https://github.com/oppia/oppia/actions/runs/3668799121/jobs/6202170482#step:8:1247

richard-johansson commented:
> @richard-johansson thanks for opening this PR! It looks like there are still a few spots left where the lint check is failing: https://github.com/oppia/oppia/actions/runs/3668799121/jobs/6202170482#step:8:1247

Ok, I will look into that. Can I test the same things locally before pushing? If so, how do I do that? Thanks @U8NWXD

vojtechjelinek commented on core/controllers/editor_test.py:
```diff
@@ -704,13 +704,16 @@ def test_exploration_download_handler_for_default_exploration(self) -> None:
-                'The title for ZIP download handler test!.yaml').read())
+        with zipfile.ZipFile(io.BytesIO(golden_zipfile)) as zf_gold:
+            # Compare saved with golden file.
+            with zf_saved.open(
+                'The title for ZIP download handler test!.yaml'
+                ) as zf_saved_open, zf_gold.open(
+                'The title for ZIP download handler test!.yaml'
+                ) as zf_gold_open:
```
```suggestion
            with zf_saved.open(
                'The title for ZIP download handler test!.yaml'
            ) as zf_saved_open, zf_gold.open(
                'The title for ZIP download handler test!.yaml'
            ) as zf_gold_open:
```

vojtechjelinek commented:
@richard-johansson You can check https://github.com/oppia/oppia/wiki/Lint-Checks on how to run the lint checks.

richard-johansson commented:
> @richard-johansson You can check https://github.com/oppia/oppia/wiki/Lint-Checks on how to run the lint checks.

Thanks, I will check that out! @vojtechjelinek

ricjo462 committed:
Added 'with' clause to fix pylint problems

ricjo462 committed:
Added with clause and removed cleanup()

ricjo462 committed:
Now using 'with' to open files, removed file close() and fixed too long lines issue.

ricjo462 committed:
Now uses 'with' to open file

ricjo462 committed:
Fixed mypy type checks. TemporaryDirectory in with clause returns name directly and no need for calling cleanup

ricjo462 committed:
Fixed pylint consider-using-with error and removed unnesecary file close

ricjo462 committed:
Fixed pylint consider-using-with error and removed unnesecary file close

ricjo462 committed:
Fixed pylint consider-using-with error, removed unnesecary file close and shortened comment line lenght

ricjo462 committed:
Fixed pylint consider-using-with error and removed unnesecary file close

ricjo462 committed:
Fixed pylint consider-using-with error

ricjo462 committed:
Fixed pylint consider-using-with error

ricjo462 committed:
Fixed pylint consider-using-with error and removed unnesecary file close

ricjo462 committed:
Fixed pylint consider-using-with error

ricjo462 committed:
Fixed pylint consider-using-with error

ricjo462 committed:
Fixed pylint consider-using-with error

ricjo462 committed:
Fixed pylint consider-using-with error

ricjo462 committed:
Fixed pylint consider-using-with error

ricjo462 committed:
Fixed pylint consider-using-with error

ricjo462 committed:
Fixed pylint consider-using-with error

ricjo462 committed:
Fixed pylint consider-using-with error

ricjo462 committed:
Fixed pylint consider-using-with error

ricjo462 committed:
Fixed pylint consider-using-with error

richard-johansson committed:
Merge branch 'develop' into pylintConsiderUsingWith

ricjo462 committed:
Merge branch 'pylintConsiderUsingWith' of github.com:richard-johansson/oppia into pylintConsiderUsingWith

ricjo462 committed:
Revert "Fixed pylint consider-using-with error"

This reverts commit 6b3fc61deda4ee369c9c57d4e0be690e43b97961.

ricjo462 committed:
Revert "Fixed pylint consider-using-with error"

This reverts commit fe30e04c5f3a693ec99db7c4a73608852a7c9dbf.

ricjo462 committed:
Revert "Fixed pylint consider-using-with error"

This reverts commit 950f5e921380bdfc2dae1658e41e6b0aabadbe08.

richard-johansson committed:
Merge branch 'develop' into pylintConsiderUsingWith

richard-johansson committed:
Fixed indentation

Co-authored-by: Vojtěch Jelínek <vojtech.jelinek@hey.com>

richard-johansson commented:
Hi @vojtechjelinek, I have some problem with the mypy type checks in several places. For example, in an attempt to  fix the pylint problem 
```
************* Module scripts.release_scripts.update_configs_test
scripts/release_scripts/update_configs_test.py:137:27: R1732: Consider using 'with' for resource-allocating operations 
```
by changing
```
def test_missing_mailgun_api_key_line(self) -> None:
    mailgun_api_key = ('key-%s' % ('').join(['1'] * 32))
    def mock_getpass(prompt: str) -> str:  # pylint: disable=unused-argument
        return mailgun_api_key
    getpass_swap = self.swap(getpass, 'getpass', mock_getpass)

    temp_feconf_path = tempfile.NamedTemporaryFile().name
    feconf_text = (
        'REDISHOST = \'192.13.2.1\'\n'
        '# When the site terms were last updated, in UTC.\n'
        'TERMS_PAGE_LAST_UPDATED_UTC = '
        'datetime.datetime(2015, 10, 14, 2, 40, 0)\n'
        '# Format of string for dashboard statistics logs.\n'
        '# NOTE TO DEVELOPERS: This format should not be changed, '
        'since it is used in\n'
        '# the existing storage models for UserStatsModel.\n'
        'DASHBOARD_STATS_DATETIME_STRING_FORMAT = \'%Y-%m-%d\'\n')
    with utils.open_file(temp_feconf_path, 'w') as f:
        f.write(feconf_text)

    with getpass_swap, self.assertRaisesRegex(
        AssertionError, 'Missing mailgun API key'
    ):
        update_configs.add_mailgun_api_key(temp_feconf_path)
```
to
```
def test_missing_mailgun_api_key_line(self) -> None:
    mailgun_api_key = ('key-%s' % ('').join(['1'] * 32))
    def mock_getpass(prompt: str) -> str:  # pylint: disable=unused-argument
        return mailgun_api_key
    getpass_swap = self.swap(getpass, 'getpass', mock_getpass)

    with tempfile.NamedTemporaryFile().name as temp_feconf_path:
        feconf_text = (
            'REDISHOST = \'192.13.2.1\'\n'
            '# When the site terms were last updated, in UTC.\n'
            'TERMS_PAGE_LAST_UPDATED_UTC = '
            'datetime.datetime(2015, 10, 14, 2, 40, 0)\n'
            '# Format of string for dashboard statistics logs.\n'
            '# NOTE TO DEVELOPERS: This format should not be changed, '
            'since it is used in\n'
            '# the existing storage models for UserStatsModel.\n'
            'DASHBOARD_STATS_DATETIME_STRING_FORMAT = \'%Y-%m-%d\'\n')
        with utils.open_file(temp_feconf_path, 'w') as f:
            f.write(feconf_text)

        with getpass_swap, self.assertRaisesRegex(
            AssertionError, 'Missing mailgun API key'
        ):
            update_configs.add_mailgun_api_key(temp_feconf_path)
```
(i.e., changing 
`temp_feconf_path = tempfile.NamedTemporaryFile().name` to `with tempfile.NamedTemporaryFile().name as temp_feconf_path`)

I get the error:
```
Starting Mypy type checks.
scripts/release_scripts/update_configs_test.py:112: error: "str" has no attribute "__enter__"; maybe "__iter__"?  [attr-defined]
scripts/release_scripts/update_configs_test.py:112: error: "str" has no attribute "__exit__"  [attr-defined]
Found 2 errors in 1 file (checked 881 source files)
```

This happens for several files, that 'str' does not have an attribute __enter__ or __exit__. Can I just add a pylint ignore for those occasions? Do you have any other suggestions on how I can the problem? Thanks

vojtechjelinek commented:
@richard-johansson You need to do `with tempfile.NamedTemporaryFile() as temp_feconf_file:` and only then `temp_feconf_name = temp_feconf_file.name`

richard-johansson commented:
> @richard-johansson You need to do `with tempfile.NamedTemporaryFile() as temp_feconf_file:` and only then `temp_feconf_name = temp_feconf_file.name`

Ok, I will test that. Thanks! @vojtechjelinek

ricjo462 committed:
Fixed consider-using-with problem and mypy checks issues

ricjo462 committed:
Now using 'with' for resource-allocating operations

ricjo462 committed:
Now using 'with' for resource-allocating operations

ricjo462 committed:
Now using 'with' to manage the acquisition and release of the lock on semaphore object.

ricjo462 committed:
Now using 'with' for resource-allocating operations

ricjo462 committed:
Now using 'with' for resource-allocating operations

ricjo462 committed:
Now using 'with' for resource-allocating operations and removed unused close()

ricjo462 committed:
Now using 'with' for resource-allocating operations

ricjo462 committed:
Now using 'with' for resource-allocating operations and removed unused close()

ricjo462 committed:
Added empty 'with' clause to correctly handle resource-allocating operations

richard-johansson commented:
@vojtechjelinek I fixed all the linting problems now and have verified it locally. However I get some error that I am unable to understand and solve. It seems that there is some js file missing and I have not touched it, can you please help me? Thanks!

https://github.com/oppia/oppia/actions/runs/3760603144/jobs/6391495652#step:8:1067

richard-johansson commented:
> @vojtechjelinek I fixed all the linting problems now and have verified it locally. However I get some error that I am unable to understand and solve. It seems that there is some js file missing and I have not touched it, can you please help me? Thanks!
> 
> https://github.com/oppia/oppia/actions/runs/3760603144/jobs/6391495652#step:8:1067

@vojtechjelinek @U8NWXD can you please take a look? Thanks

richard-johansson committed:
Merge branch 'develop' into pylintConsiderUsingWith

vojtechjelinek commented on scripts/linters/js_ts_linter.py:
```diff
@@ -141,11 +141,11 @@ def compile_all_ts_files() -> None:
-        cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, shell=True)
-
-    _, encoded_stderr = proc.communicate()
+        './tsconfig.json', COMPILED_TYPESCRIPT_TMP_PATH)
```
Unexpected change on this line.

richard-johansson replied:
I have changed that line back to what it was before now. https://github.com/oppia/oppia/pull/16701/commits/ed6ceda9ebe7eb802fd0e28ad3fb2df03cae9265

vojtechjelinek commented:
> > @vojtechjelinek I fixed all the linting problems now and have verified it locally. However I get some error that I am unable to understand and solve. It seems that there is some js file missing and I have not touched it, can you please help me? Thanks!
> > [oppia/oppia/actions/runs/3760603144/jobs/6391495652#step:8:1067](https://github.com/oppia/oppia/actions/runs/3760603144/jobs/6391495652#step:8:1067)
> 
> @vojtechjelinek @U8NWXD can you please take a look? Thanks

See https://github.com/oppia/oppia/pull/16701/files#r1057882945

Also, the backend tests are failing, that is because we swap `Popen` to a normal function and not a context manager, let me know if you need help with that.

richard-johansson committed:
Fixed unexpected change

richard-johansson commented:
> 



> > > @vojtechjelinek I fixed all the linting problems now and have verified it locally. However I get some error that I am unable to understand and solve. It seems that there is some js file missing and I have not touched it, can you please help me? Thanks!
> > > [oppia/oppia/actions/runs/3760603144/jobs/6391495652#step:8:1067](https://github.com/oppia/oppia/actions/runs/3760603144/jobs/6391495652#step:8:1067)
> > 
> > 
> > @vojtechjelinek @U8NWXD can you please take a look? Thanks
> 
> See https://github.com/oppia/oppia/pull/16701/files#r1057882945
> 
> Also, the backend tests are failing, that is because we swap `Popen` to a normal function and not a context manager, let me know if you need help with that.

Yes please give me some help on that. @vojtechjelinek

richard-johansson committed:
Merge branch 'develop' into pylintConsiderUsingWith

U8NWXD commented:
Deassigning myself since I'm not reviewing PRs right now

vojtechjelinek commented:
> > 
> 
> > > > @vojtechjelinek I fixed all the linting problems now and have verified it locally. However I get some error that I am unable to understand and solve. It seems that there is some js file missing and I have not touched it, can you please help me? Thanks!
> > > > [oppia/oppia/actions/runs/3760603144/jobs/6391495652#step:8:1067](https://github.com/oppia/oppia/actions/runs/3760603144/jobs/6391495652#step:8:1067)
> > > 
> > > 
> > > @vojtechjelinek @U8NWXD can you please take a look? Thanks
> > 
> > 
> > See [#16701 (files)](https://github.com/oppia/oppia/pull/16701/files#r1057882945)
> > Also, the backend tests are failing, that is because we swap `Popen` to a normal function and not a context manager, let me know if you need help with that.
> 
> Yes please give me some help on that. @vojtechjelinek

In all places where we use `self.swap(subprocess, 'Popen', some_function)` you need to go to the place where we define the `some_function` and then add a `@contextlib.contextmanager` decorator, that should fix the issue.

richard-johansson committed:
Merge branch 'develop' into pylintConsiderUsingWith

ricjo462 committed:
Added contextlib decorator since 'Popen' is swapped to a normal function and not a context manager anymore

richard-johansson commented:
> > > > > @vojtechjelinek I fixed all the linting problems now and have verified it locally. However I get some error that I am unable to understand and solve. It seems that there is some js file missing and I have not touched it, can you please help me? Thanks!
> > > > > [oppia/oppia/actions/runs/3760603144/jobs/6391495652#step:8:1067](https://github.com/oppia/oppia/actions/runs/3760603144/jobs/6391495652#step:8:1067)
> > > > 
> > > > 
> > > > @vojtechjelinek @U8NWXD can you please take a look? Thanks
> > > 
> > > 
> > > See [#16701 (files)](https://github.com/oppia/oppia/pull/16701/files#r1057882945)
> > > Also, the backend tests are failing, that is because we swap `Popen` to a normal function and not a context manager, let me know if you need help with that.
> > 
> > 
> > Yes please give me some help on that. @vojtechjelinek
> 
> In all places where we use `self.swap(subprocess, 'Popen', some_function)` you need to go to the place where we define the `some_function` and then add a `@contextlib.contextmanager` decorator, that should fix the issue.

Hello @vojtechjelinek! Doing as you suggested did not work, I now fail the mypy checks with `scripts/typescript_checks_test.py:107: error: Argument 1 to "contextmanager" has incompatible type "Callable[[str, str, str], Popen[str]]"; expected "Callable[..., Iterator[<nothing>]]"  [arg-type]`

I have added the context manager in `typescript_checks_test.py` to start with
```
def test_error_is_raised_for_invalid_compilation_of_tsconfig(self) -> None:
    """Test that error is produced if stdout is not empty."""
    with subprocess.Popen(
        ['echo', 'test'], stdout=subprocess.PIPE,
        encoding='utf-8') as process:
        @contextlib.contextmanager
        def mock_popen_for_errors(
            unused_cmd: str, stdout: str, encoding: str  # pylint: disable=unused-argument
        ) -> subprocess.Popen[str]:  # pylint: disable=unsubscriptable-object
            return process

    with self.swap(subprocess, 'Popen', mock_popen_for_errors):
        with self.assertRaisesRegex(SystemExit, '1'):
            typescript_checks.compile_and_check_typescript(
                typescript_checks.TSCONFIG_FILEPATH)
```

can you help me?

vojtechjelinek commented:
@richard-johansson I think replacing `return process` with `yield process` should fix this.

richard-johansson commented:
> @richard-johansson I think replacing `return process` with `yield process` should fix this.

@vojtechjelinek unfortunately that results in mypy checks failing with two errors:
```
scripts/typescript_checks_test.py:107: error: Argument 1 to "contextmanager" has incompatible type "Callable[[str, str, str], Popen[str]]"; expected "Callable[..., Iterator[<nothing>]]"  [arg-type]
scripts/typescript_checks_test.py:108: error: The return type of a generator function should be "Generator" or one of its supertypes  [misc]
Found 2 errors in 1 file (checked 883 source files)
```

vojtechjelinek commented:
@richard-johansson Hey, this should work

```python
    def test_error_is_raised_for_invalid_compilation_of_tsconfig(self) -> None:
        """Test that error is produced if stdout is not empty."""

        with subprocess.Popen(
            ['echo', 'test'], stdout=subprocess.PIPE, encoding='utf-8'
        ) as process:
            @contextlib.contextmanager
            def mock_popen_for_errors(
                unused_cmd: str, stdout: str, encoding: str  # pylint: disable=unused-argument
            ) -> Iterator[subprocess.Popen[str]]:  # pylint: disable=unsubscriptable-object
                yield process

            with self.swap(subprocess, 'Popen', mock_popen_for_errors):
                with self.assertRaisesRegex(SystemExit, '1'):
                    typescript_checks.compile_and_check_typescript(
                        typescript_checks.TSCONFIG_FILEPATH)
```

ricjo462 committed:
In all places where `self.swap(subprocess, 'Popen', some_function)` is used, in `some_function` a @contextlib.contextmanager decorator is added. Replacing return with yield and changing return type to Iterator[] in some places.

ricjo462 committed:
import formatting linter error

ricjo462 committed:
Fixed linter error, break after parenthesis

richard-johansson commented:
> @richard-johansson Hey, this should work
> 
> ```python
>     def test_error_is_raised_for_invalid_compilation_of_tsconfig(self) -> None:
>         """Test that error is produced if stdout is not empty."""
> 
>         with subprocess.Popen(
>             ['echo', 'test'], stdout=subprocess.PIPE, encoding='utf-8'
>         ) as process:
>             @contextlib.contextmanager
>             def mock_popen_for_errors(
>                 unused_cmd: str, stdout: str, encoding: str  # pylint: disable=unused-argument
>             ) -> Iterator[subprocess.Popen[str]]:  # pylint: disable=unsubscriptable-object
>                 yield process
> 
>             with self.swap(subprocess, 'Popen', mock_popen_for_errors):
>                 with self.assertRaisesRegex(SystemExit, '1'):
>                     typescript_checks.compile_and_check_typescript(
>                         typescript_checks.TSCONFIG_FILEPATH)
> ```

@vojtechjelinek I fixed that and the mypy checks passed. Now after pushing there seems to be more errors, can you take a look at what I should fix?

richard-johansson committed:
Merge branch 'develop' into pylintConsiderUsingWith

vojtechjelinek commented on scripts/extend_index_yaml_test.py:
```diff
@@ -30,8 +30,10 @@ class ExtendIndexYamlTest(test_utils.GenericTestBase):
         super().setUp()
-        self.index_yaml_file = tempfile.NamedTemporaryFile()
-        self.web_inf_index_yaml_file = tempfile.NamedTemporaryFile()
+        with tempfile.NamedTemporaryFile() as index_yaml_file:
+            self.index_yaml_file = index_yaml_file
+        with tempfile.NamedTemporaryFile() as web_inf_index_yaml_file:
+            self.web_inf_index_yaml_file = web_inf_index_yaml_file
```
As these two are only used in `_run_test_for_extend_index_yaml` (through `self.index_yaml_file_name` and `self.index_yaml_swap`) you can just move them there. That should help with the issue you're facing.

richard-johansson replied:
Do you mean that I should move all code from `setup()` to `_run_test_for_extend_index_yaml()`? Because `index_yaml_file` and `web_inf_index_yaml_file` are used in many places in the `setup()` function. @vojtechjelinek

vojtechjelinek replied:
Not sure if all, but if the variables that depend on `index_yaml_file` or `web_inf_index_yaml_file` probably yes. You will have to try it out.

richard-johansson replied:
Hi @vojtechjelinek, when changing as you suggested the linter checks fail with error:

```
************* Module scripts.extend_index_yaml_test
scripts/extend_index_yaml_test.py:31:4: W0235: Useless super delegation in method 'setUp' 
...
scripts/extend_index_yaml_test.py:38:12: W0201: Attribute 'index_yaml_file' defined outside __init__ 
scripts/extend_index_yaml_test.py:40:12: W0201: Attribute 'web_inf_index_yaml_file' defined outside __init__ 
```

vojtechjelinek replied:
@richard-johansson I do not now the exact code that you have right now. Please send it too. Also, did you try to search for these problems online and solve them?

vojtechjelinek commented:
@richard-johansson https://github.com/oppia/oppia/pull/16701/files#r1063975945

U8NWXD commented:
Deassigning myself for now since PR is still a WIP. I can do a codeowner review once vojtechjelinek approves (unless there's something you want me in particular to look at, in which case I'm happy to help)

richard-johansson committed:
Merge branch 'develop' into pylintConsiderUsingWith

richard-johansson committed:
Merge branch 'develop' into pylintConsiderUsingWith
